adminSiteServer/apiRouter.ts

```diff
@@ -3051,6 +3059,169 @@ postRouteWithRWTransaction(
     }
 )
 
+getRouteNonIdempotentWithRWTransaction(
+    apiRouter,
+    "/images.json",
+    async (_, res, trx) => {
+        try {
+            const images = await db.getCloudflareImages(trx)
+            res.set("Cache-Control", "no-store")
+            res.send({ images })
+        } catch (error) {
+            console.error("Error fetching images", error)
+            res.status(500).json({
+                error: { message: String(error), status: 500 },
+            })
+        }
+    }
+)
+
+postRouteWithRWTransaction(apiRouter, "/image", async (req, res, trx) => {
+    const { filename, type, content } = req.body
+    if (!filename || !type || !content) {
+        throw new JsonError("Missing required fields", 400)
+    }
+    if (
+        typeof filename !== "string" ||
+        typeof type !== "string" ||
+        typeof content !== "string"
+    ) {
+        throw new JsonError("Invalid field types", 400)
+    }
+
+    // Strip the data URL prefix
+    const stripped = content.slice(content.indexOf(",") + 1)
+    const asBuffer = Buffer.from(stripped, "base64")
+    const asBlob = new Blob([asBuffer], { type: type })
+    const dimensions = await sharp(asBuffer)
+        .metadata()
+        .then(({ width, height }) => ({ width, height }))
+
+    const body = new FormData()
+    body.append("file", asBlob)
+    body.append("id", encodeURIComponent(filename))
+    body.append("metadata", JSON.stringify({ filename }))
+    body.append("requireSignedURLs", "false")
+
+    console.log("Uploading image:", filename)
+    const response = await fetch(
+        `https://api.cloudflare.com/client/v4/accounts/${CLOUDFLARE_IMAGES_ACCOUNT_ID}/images/v1`,
+        {
+            method: "POST",
+            headers: {
+                Authorization: `Bearer ${CLOUDFLARE_IMAGES_API_KEY}`,
+            },
+            body,
+        }
+    ).then((res) => res.json())
+
+    if (response.errors.length) {
+        console.error("Error uploading image to Cloudflare:", response.errors)
+        throw new JsonError(JSON.stringify(response.errors))
+    }
+
+    const cloudflareId = response.result.id
+
+    if (!cloudflareId) {
+        return {
+            success: false,
+            error: "Failed to upload image",
+        }
+    }
+
+    await trx<DbEnrichedImage>("images").insert({
+        filename,
+        originalWidth: dimensions.width,
+        originalHeight: dimensions.height,
+        cloudflareId,
+        // TODO: make defaultAlt nullable
+        defaultAlt: "Default alt text",
+        // TODO: drop googleId
+        googleId: String(Math.random()).slice(2),
```

Comment:
We're gonna drop it soon, but you're right that I should just make it nullable in the migration so that we don't have to risk an improbable collision here for no reason.

Reply:
Cool, thanks! I was mostly worried about getting bogus data when asking questions like "what is the average length of our alt texts" if we kept the dummy values around.